server/src/transport/ws.rs

```diff
@@ -0,0 +1,464 @@
+	pub(crate) max_response_body_size: u32,
+	pub(crate) max_log_length: u32,
+	pub(crate) resources: &'a Resources,
+	pub(crate) sink: &'a MethodSink,
+	pub(crate) logger: &'a L,
+	pub(crate) request_start: L::Instant,
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct Call<'a, L: WsLogger> {
+	pub(crate) params: Params<'a>,
+	pub(crate) name: &'a str,
+	pub(crate) call: CallData<'a, L>,
+	pub(crate) id: Id<'a>,
+}
+
+// Batch responses must be sent back as a single message so we read the results from each
+// request in the batch and read the results off of a new channel, `rx_batch`, and then send the
+// complete batch response back to the client over `tx`.
+pub(crate) async fn process_batch_request<L: WsLogger>(b: Batch<'_, L>) -> BatchResponse {
+	let Batch { data, call } = b;
+
+	if let Ok(batch) = serde_json::from_slice::<Vec<Request>>(&data) {
+		return if !batch.is_empty() {
+			let batch = batch.into_iter().map(|req| Ok((req, call.clone())));
+			let batch_stream = futures_util::stream::iter(batch);
+
+			let trace = RpcTracing::batch();
+
+			return async {
+				let max_response_size = call.max_response_body_size;
+
+				let batch_response = batch_stream
+					.try_fold(
+						BatchResponseBuilder::new_with_limit(max_response_size as usize),
+						|batch_response, (req, call)| async move {
+							let params = Params::new(req.params.map(|params| params.get()));
+							let response = execute_call(Call { name: &req.method, params, id: req.id, call }).await;
+							batch_response.append(response.as_inner())
+						},
+					)
+					.await;
+
+				match batch_response {
+					Ok(batch) => batch.finish(),
+					Err(batch_err) => batch_err,
+				}
+			}
+			.instrument(trace.into_span())
+			.await;
+		} else {
+			BatchResponse::error(Id::Null, ErrorObject::from(ErrorCode::InvalidRequest))
+		};
+	}
+
+	let (id, code) = prepare_error(&data);
+	BatchResponse::error(id, ErrorObject::from(code))
+}
+
+pub(crate) async fn process_single_request<L: WsLogger>(data: Vec<u8>, call: CallData<'_, L>) -> MethodResult {
+	if let Ok(req) = serde_json::from_slice::<Request>(&data) {
+		let trace = RpcTracing::method_call(&req.method);
+
+		async {
+			rx_log_from_json(&req, call.max_log_length);
+
+			let params = Params::new(req.params.map(|params| params.get()));
+			let name = &req.method;
+			let id = req.id;
+
+			execute_call(Call { name, params, id, call }).await
+		}
+		.instrument(trace.into_span())
+		.await
+	} else {
+		let (id, code) = prepare_error(&data);
+		MethodResult::SendAndLogger(MethodResponse::error(id, ErrorObject::from(code)))
+	}
+}
+
+/// Execute a call which returns result of the call with a additional sink
+/// to fire a signal once the subscription call has been answered.
+///
+/// Returns `(MethodResponse, None)` on every call that isn't a subscription
+/// Otherwise `(MethodResponse, Some(PendingSubscriptionCallTx)`.
+pub(crate) async fn execute_call<L: WsLogger>(c: Call<'_, L>) -> MethodResult {
```

Comment:
How much similarity is there between these methods and the same in `http.rs`; is it worth trying to reuse more functionality from within them or are the yactually fairly distinct?

Reply:
Makes sense to me; definitely no rush, just something I pondered when I saw it!